plugins/autoslab.cpp

```diff
@@ -0,0 +1,267 @@
+static std::string get_first_name(df::unit *unit)
+{
+    return Translation::capitalize(unit->name.first_name);
+}
+
+static std::string get_last_name(df::unit *unit)
+{
+    df::language_name name = unit->name;
+    std::string ret = "";
+    for (int i = 0; i < 2; i++)
+    {
+        if (name.words[i] >= 0)
+            ret += *world->raws.language.translations[name.language]->words[name.words[i]];
+    }
+    return Translation::capitalize(ret);
+}
+
+// Couldn't figure out any other way to do this besides look for the dwarf name in
+// the slab item description.
+// Ideally, we could get the historical figure id from the slab but I didn't
+// see anything like that in the item struct. This seems to work based on testing.
+// Confirmed nicknames don't show up in engraved slab names, so this should probably work okay
+bool engravedSlabItemExists(df::unit *unit, std::vector<df::item *> slabs)
+{
+    for (auto slab : slabs)
+    {
+        std::string desc = "";
+        slab->getItemDescription(&desc, 0);
+        auto fullName = get_first_name(unit) + " " + get_last_name(unit);
+        if (desc.find(fullName) != std::string::npos)
+            return true;
+    }
+
+    return false;
+}
+
+// Queue up a single order to engrave the slab for the given unit
+static void createSlabJob(df::unit *unit)
+{
+    auto next_id = world->manager_order_next_id++;
+    auto order = new df::manager_order();
+
+    order->id = next_id;
+    order->job_type = df::job_type::EngraveSlab;
+    order->hist_figure_id = unit->hist_figure_id;
+    order->amount_left = 1;
+    order->amount_total = 1;
+    world->manager_orders.push_back(order);
+}
+
+static void checkslabs(color_ostream &out)
+{
+    // Get existing orders for slab engraving as map hist_figure_id -> order ID
+    std::map<int32_t, int32_t> histToJob;
+    for (auto order : world->manager_orders)
+    {
+        if (order->job_type == df::job_type::EngraveSlab)
+            histToJob[order->hist_figure_id] = order->id;
+    }
+
+    // Get list of engraved slab items on map
+    std::vector<df::item *> engravedSlabs;
+    std::copy_if(world->items.all.begin(), world->items.all.end(),
+                 std::back_inserter(engravedSlabs),
+                 [](df::item *item)
+                 { return item->getType() == df::item_type::SLAB && item->getSlabEngravingType() == df::slab_engraving_type::Memorial; });
+
+    // Build list of ghosts
+    std::vector<df::unit *> ghosts;
+    std::copy_if(world->units.all.begin(), world->units.all.end(),
+                 std::back_inserter(ghosts),
+                 [](const auto &unit)
+                 { return unit->flags3.bits.ghostly; });
+
+    for (auto ghost : ghosts)
+    {
+        // Only create a job is the map has no existing jobs for that historical figure or no existing engraved slabs
+        if (histToJob.count(ghost->hist_figure_id) == 0 && !engravedSlabItemExists(ghost, engravedSlabs))
+        {
+            createSlabJob(ghost);
```

Comment:
if there exist no free unengraved slabs, you might also want to create a job for an unengraved slab to use for the engrave job.